allennlp/data/iterators/bucket_iterator.py

```diff
@@ -56,35 +56,40 @@ def __init__(self,
                  sorting_keys: List[Tuple[str, str]] = None,
                  padding_noise: float = 0.1,
                  biggest_batch_first: bool = False,
-                 batch_size: int = 32) -> None:
+                 batch_size: int = 32,
+                 instances_per_epoch: int = None,
+                 max_instances_in_memory: int = None) -> None:
         self._sorting_keys = sorting_keys or []
         self._padding_noise = padding_noise
         self._biggest_batch_first = biggest_batch_first
-        super(BucketIterator, self).__init__(batch_size)
+        super(BucketIterator, self).__init__(batch_size=batch_size,
+                                             instances_per_epoch=instances_per_epoch,
+                                             max_instances_in_memory=max_instances_in_memory)
 
     @overrides
     def _create_batches(self, instances: Iterable[Instance], shuffle: bool) -> Iterable[Batch]:
-        instances = ensure_list(instances)
+        for instance_list in self._memory_sized_lists(instances):
+            if self._sorting_keys:
```

Comment:
I'm trying to think now why you would ever want a bucket iterator that doesn't have sorting keys...  I think we can remove this check, and make the argument required in the constructor and in `from_params`.  It doesn't make sense to bucket unless you give a means to sort, that's kind of the definition of the difference between the "basic" and the "bucket" iterators.  @DeNeutoy, do you agree?  Am I missing something?

Reply:
Yeah sorting keys should be required.

Reply:
I'll leave this for a future PR

Reply:
Seems like it would be quick, and very important - if you don't specify this, your model could train an order of magnitude slower

Reply:
oops, hit merge too fast, I'll do another really quick PR